### ISSUE TITLE
gem_package behaves like chef_gem after chef installs system ruby

### ISSUE THREAD

**Issue by duallain** [chef/chef#3363]
I expect that this recipe when run for the first time will install the json gem to /usr/bin/gem:

``` ruby
package 'rubygems' do
  action :install
end

gem_package 'json' do
  action :install
  gem_binary '/usr/bin/gem'
end

gem_package 'thor' do
  action :install
  gem_binary '/usr/bin/gem'
end

execute "/usr/bin/gem list |grep thor"
execute "/usr/bin/gem list |grep json"
```

thor is installed to /usr/bin/gem but json is not.  This is for chef 12.3.0 in test kitchen.

The json gem is not installed because of this message:

```
  * gem_package[json] action install[2015-05-11T22:07:19+00:00] INFO: Processing gem_package[json] action install (repo::default line 16)       
 (up to date)       
  * gem_package[thor] action install[2015-05-11T22:07:24+00:00] INFO: Processing gem_package[thor] action install (repo::default line 21)       
[2015-05-11T22:07:30+00:00] INFO: gem_package[thor] installed thor at 0.19.1
```

Testing with chef 11.16.4 the recipe runs all the way though without issue.

```
* gem_package[json] action install[2015-05-11T22:10:59+00:00] INFO: Processing gem_package[json] action install (repo::default line 16)      
- install version 1.8.2 of package json
```

**Comment by lamont-granquist** [comment]
You most likely need to use `action :upgrade`.  The json gem is a default gem in ruby:

```
# gem uninstall json
ERROR:  While executing gem ... (Gem::InstallError)
    gem "json" cannot be uninstalled because it is a default gem
```

If you try to `:install` it then it will find it and won't install it.

If you try to `:upgrade` it then it'll download the latest if its available:

```
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * gem_package[json] action upgrade
    - upgrade package json from 1.8.1 to 1.8.2
```

I suspect between 11.16.4 and 12.3.0 that we may have fixed this.

**Comment by duallain** [comment]
That seemed to work, but I'm not sure that it is the full/correct solution.

The json gem is installed in /opt/embedded/chef. But is not installed for /usr/bin/gem. Should the fact that  a gem is installed in one path affect if it is installed in a different one?  Especially with the difference between chef_gem and gem_package.

I'm now fixed, but it is possible that this is a bigger problem.

**Comment by lamont-granquist** [comment]
What does this do before you run chef on the box?  And what version of ruby are you running?

```ruby
% /usr/bin/irb
2.1.5 :001 > require 'json'
```

**Comment by duallain** [comment]
If I converge the image without installing ruby gems:
```
$ /usr/bin/irb
-bash: /usr/bin/irb: No such file or directory
```
At the same point
```
$ /opt/chef/embedded/bin/irb 
irb(main):001:0> require 'json'
=> true
```

after installing rubygems

```
$ /usr/bin/irb
irb(main):001:0> require 'json'
LoadError: no such file to load -- json
```
the chef embedded hasn't changed.

This is what the gem list for the ubuntu installation of rubygems gives you:
```
$ /usr/bin/gem list

*** LOCAL GEMS ***

```

I then add this block to the run:
```
gem_package 'json' do
  action :install
  gem_binary '/usr/bin/gem'
end
```

And json is not installed because:

```
  * gem_package[json] action install[2015-05-13T15:38:15+00:00] INFO: Processing gem_package[json] action install (repo::default line 16)       
 (up to date)
```

```
$ /usr/bin/gem list

*** LOCAL GEMS ***
```
But when following the same pattern with a different gem

Before:
```
$ /opt/chef/embedded/bin/irb
irb(main):001:0> require 'thor'
LoadError: cannot load such file -- thor
```

The block:
```
gem_package 'thor' do
  action :install
  gem_binary '/usr/bin/gem'
end
```

After:

```
$ /usr/bin/gem list

*** LOCAL GEMS ***
thor (0.19.1)
```

Why does a gem being installed in /opt/chef/embedded/ matter to if a gem should be installed to a different gem location.  Particularly when the gem_binary is explicitly set.

More than that, any gem that is installed in /opt/chef/embedded/ acts the same way.
Before:
```
$ /opt/chef/embedded/bin/irb
irb(main):001:0> require 'pry'
=> true
```

```
$ /usr/bin/irb
irb(main):001:0> require 'pry'
LoadError: no such file to load -- pry
```

this block gets ran:
```
gem_package 'pry' do
  action :install
  gem_binary '/usr/bin/gem'
end
```

and the require 'pry' for /usr/bin/irb fails the same way.

**Comment by lamont-granquist** [comment]
okay, i think i'm parsing what your problem is at least.  and your version of /usr/bin/ruby is sufficiently old that the json gem hasn't been integrated into ruby yet?

what happens if you leave "gem_binary" off of the gem_package resource?

**Comment by lamont-granquist** [comment]
I tried to replicate this with chef 12.3.0 and the 'highline' gem as my installed-in-embedded-but-not-in-system gem and its working for me.  Both with and without the 'gem_package' resource.  The fact that its installed in embedded doesn't affect it being installed into the system gems.

**Comment by lamont-granquist** [comment]
There is a known bug where gem_package will behave like chef_gem if there is no system gem, so if you've got an ordering issue and you're doing a gem_package before you've installed ruby+rubygems on the system then it'll behave this way.  Other than that my suspicion is some kind of pathology around the system ruby or rubygems being very old?

**Comment by duallain** [comment]
By the way, thank you so much for all your help.  It feels like we're quite close to getting it solved.

The repo recipe above is getting ruby installed by doing the package installation of rubygems.  It seems like maybe the issue is related to GEM_HOME or other ruby path/environment variables.

To be honest we don't care too much about solving the particular recipe above.  We'd be perfectly happy with any way to:

1) install ruby in a recipe
2) install json (and other gems)
3) use that ruby and gems in some scripts

 Is there a recommended/supported way of installing the system ruby?  I assume that's the place we're falling down currently.  Poise-ruby looks like it is the recommended solution.  We're using ubuntu 12.04 and ubuntu 14.04

**Comment by lamont-granquist** [comment]
Yeah it might be that we're caching information somewhere in a chef_gem or gem_package call which is done prior to your installation of the system ruby.  Then after you've installed the system ruby the rubygems caching in the provider is still hanging around and chef still 'thinks' that the embedded ruby is the only ruby on the system.  Then you're hitting that issue where gem_package falls back to chef_gem behavior if there's no ruby on the system even though you've installed it.  This is very plausible because there's both a lot of caching of information in gem_package and there's that known issue.

So, one simple test is if you've manually installed the system ruby before running chef-client does that fix its behavior and the gem_package correctly installs into the system?

And best practices are all over the place.  If you don't want to spend the time compiling ruby, I'd just install the packages you need directly via the package resource and avoid any cookbooks out there.  If you want to have an upgraded ruby on the system and/or manage multiple ones I'd suggest the rbenv or chruby cookbooks (rbenv is a bit heavier than chruby, but the rbenv cookbook seems to be getting more maintenance than the chruby cookbook -- although if the chruby one works the both chruby and its cookbook may be simplest solution overall).

**Comment by lamont-granquist** [comment]
Did you ever determine if manually installing the system ruby before running chef-client fixed the behavior (and therefore points at some caching going on?)

**Comment by lamont-granquist** [comment]
There are two things that we could possibly do to fix this:

- make gem_package raise if there's no system ruby, instead of behaving like chef_gem
- track down whatever internal state is getting cached that doesn't notice the system ruby getting installed

**Comment by dcampano** [comment]
We are also seeing this same issue when trying to install a gem to the system ruby.  The gem (io-console) exists in chef embedded and therefore it is not being installed to the system ruby.

```
gem_package "io-console" do
  gem_binary "/usr/bin/gem"
end
```

Chef-client debug reports that the gem already exists:

[2015-09-02T10:01:02-05:00] DEBUG: gem_package[io-console] using gem '/usr/bin/gem'
[2015-09-02T10:01:02-05:00] DEBUG: gem_package[io-console] found installed gem io-console version 0.4.3 matching io-console (>= 0)
[2015-09-02T10:01:02-05:00] DEBUG:  found gem io-console version 0.4.2 for platform ruby from #<Gem::Source:0x0000000682fb30>
[2015-09-02T10:01:02-05:00] DEBUG: gem_package[io-console] io-console 0.4.3 already installed
[2015-09-02T10:01:02-05:00] DEBUG: gem_package[io-console] is already installed - nothing to do
 (up to date)

When we bootstrap our systems, the first thing that we do is install ruby and rubygems to the system.  After that we grab the embedded chef package and install that.

Here are the system ruby and rubygems versions:

[root@ip-10-0-*-* ~]# /usr/bin/ruby -v
ruby 2.0.0p645 (2015-04-13) [x86_64-linux]
[root@ip-10-0-*-* ~]# /usr/bin/gem -v
2.4.8

Our chef-client version is:

Chef: 12.4.1

**Comment by lamont-granquist** [comment]
So we still need to know if you've already installed the ruby+rubygems to the system before the chef-client run, does it work?

**Comment by dcampano** [comment]
When we bootstrap our systems, the first thing that we do is install ruby+rubygems to the system. After that we grab the embedded chef package and install that.

We are running on Amazon Linux (2014.09) and we run the following as the first line in our bootstrap process:

yum install -q -y ruby ruby-devel gcc gcc-c++ automake autoconf rubygems make

Then we install chef and run chef-client for the first time.

**Comment by lamont-granquist** [comment]
Okay that's useful to know.  Does it work if you leave off the gem_binary flag to the gem_package provider or not?  (You shouldn't need to do that).

**Comment by dcampano** [comment]
I removed gem_binary flag and it still behaves the same.

**Comment by lamont-granquist** [comment]
Well, that's a substantially worse bug, but I also can't seem to replicate it.

**Comment by lamont-granquist** [comment]
ahhh, do you have `/opt/chef/embedded/bin` and/or `/opt/chef/bin` in your $PATH?

**Comment by lamont-granquist** [comment]
(or the chefdk versions if you're pulling chef-client out of chefdk)

**Comment by dcampano** [comment]
My path appears to be correct.  My installation is in /opt/chef/embedded but it does not show up in the path.

[root@ip-10-0-*-* ~]# echo $PATH
/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/aws/bin:/root/bin

Is there any other debugging that I can do on my end for you?  I'm not too familiar with the source code but is there a spot that you could have me edit to spit out what the current path is when running gem_package?

**Comment by lamont-granquist** [comment]
I think you can just:

```
ruby_block do
  block do
    puts ENV['PATH']
  end
end
```

put that right before the gem_package resource.

**Comment by dcampano** [comment]
@lamont-granquist Here is the output from running that.  My system version of "gem" exists in /usr/bin and that path is being hit before /opt/chef/embedded/bin

[2015-09-02T13:50:04-05:00] DEBUG: Providers for generic ruby_block resource enabled on node include: [Chef::Provider::RubyBlock]
[2015-09-02T13:50:04-05:00] DEBUG: Provider for action run on resource ruby_block[output_path] is Chef::Provider::RubyBlock
/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/aws/bin:/root/bin:/opt/chef/embedded/bin:/opt/chef/embedded/bin
[2015-09-02T13:50:04-05:00] INFO: ruby_block[output_path] called

**Comment by lamont-granquist** [comment]
Yeah I think chef-client adds those to the end of the PATH.  I can only replicate this behavior though if I add those to the start of the PATH.

**Comment by dcampano** [comment]
@lamont-granquist 

Trying to run through the code and figure out what is going on.  I noticed that "new_version_array" property in lib/chef/provider/package.rb contains ["0.4.3"] on my machine and this is the version of io-console installed in /opt/chef/embedded, but the version that would install with a /usr/bin/gem install io-console is version 0.4.2.  

What system version of Ruby are you using to try to reproduce?

**Comment by kplimack** [comment]
@dcampano @lamont-granquist I'm also running into the on OEL6.5. 
# ruby 1.8.7 (2013-06-27 patchlevel 374) [x86_64-linux]
# gem -v 1.3.7

**Comment by mrballcb** [comment]
I am experiencing the same behavior:

```
  %w{ json bigdecimal mysql2 }.each do |pkg|
    gem_package pkg do
      gem_binary '/usr/bin/gem'
      action :install
    end
  end
```

Before I run chef-client:

```
[root@mmbr ~]# gem list | grep -E 'mysql2|json|bigdecimal'
json (1.5.5)
[root@mmbr ~]# /opt/chef/embedded/bin/gem list | grep -E 'mysql2|json|bigdecimal'
bigdecimal (1.2.4)
json (1.8.3, 1.8.1)
multi_json (1.11.2)
```

Chef is run and reports:
```
  * gem_package[json] action install (up to date)
  * gem_package[bigdecimal] action install (up to date)
  * gem_package[mysql2] action install
    - install version 0.4.4 of package mysql2
```
Afterward:
```
[root@mmbr ~]# gem list | grep -E 'mysql2|json|bigdecimal'
json (1.5.5)
mysql2 (0.4.4)
[root@mmbr ~]# /opt/chef/embedded/bin/gem list | grep -E 'mysql2|json|bigdecimal'
bigdecimal (1.2.4)
json (1.8.3, 1.8.1)
multi_json (1.11.2)
```

And the reason why I need it to exist in the system gems:
```
# ./apiuser.rb
/usr/share/rubygems/rubygems/custom_require.rb:36:in `require': cannot load such file -- bigdecimal (LoadError)
	from /usr/share/rubygems/rubygems/custom_require.rb:36:in `require'
	from /usr/local/share/gems/gems/mysql2-0.4.4/lib/mysql2.rb:3:in `<top (required)>'
	from /usr/share/rubygems/rubygems/custom_require.rb:60:in `require'
	from /usr/share/rubygems/rubygems/custom_require.rb:60:in `rescue in require'
	from /usr/share/rubygems/rubygems/custom_require.rb:35:in `require'
	from /etc/monitor/scan.d/apiuser/lib/mysql2json.rb:1:in `<top (required)>'
	from ./apiuser.rb:2:in `require_relative'
	from ./apiuser.rb:2:in `<main>'
```

These systems are CentOS 6.6 in AWS w/ chef 12.6.0 client.  Prior to this code snippet, there is a mysql recipe which does some chef_gem magic.  That is probably where the chef embedded gem set is getting cached?  Is there any way to clear it?

For the time being I am just doing it manually:
```
  %w{ json bigdecimal mysql2 }.each do |pkg|
    # Chef bug 3363: gem_package acts like chef_gem after previous runs
    #gem_package pkg do
    #  gem_binary '/usr/bin/gem'
    #  action :install
    #end

    execute "Install gem #{pkg}" do
      command "/usr/bin/gem install #{pkg}"
      only_if  "[ -z $(/usr/bin/gem list | grep -E '^#{pkg} ') ]"
    end
  end
```

**Comment by kadamwolfe** [comment]
I've ran into this issue as well.  running chef clients 12.5 and 12.12.15 on OEL 6.6 and 6.7.  It might be worth noting that I do not have this issue with CentOS 7.2 (chef-client 12.12.15).

**Comment by tophercullen** [comment]
I'm seeing this issue as well. It appears to be something to do with amazon linux. When I ran the same recipe on RHEL7, it works. To add more confusion, its not all gems. For example, this code

`gem_package 'bundler'
gem_package 'io-console'`

Against this list of gems:

`gem list --local

*** LOCAL GEMS ***

bigdecimal (1.2.0)
json (1.8.3)
psych (2.0.0)
rdoc (4.2.2)
`
Does this:

`[2017-01-18T18:20:29+00:00] INFO: Processing gem_package[bundler] action install (newrelic_plugins::aws_cloudwatch line 52)
[2017-01-18T18:20:29+00:00] DEBUG: Providers for generic gem_package resource enabled on node include: [Chef::Provider::Package::Rubygems]
[2017-01-18T18:20:29+00:00] DEBUG: Provider for action install on resource gem_package[bundler] is Chef::Provider::Package::Rubygems
[2017-01-18T18:20:29+00:00] DEBUG: gem_package[bundler] detected omnibus installation in /opt/chef/embedded/bin
[2017-01-18T18:20:29+00:00] DEBUG: gem_package[bundler] searching for 'gem' binary in path: /sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin:/opt/chef/embedded/bin:/opt/chef/embedded/bin:/usr/local/sbin:/usr/local/bin
[2017-01-18T18:20:29+00:00] DEBUG: gem_package[bundler] using gem '/bin/gem'
/root/.gem/ruby/2.0:/usr/share/ruby/gems/2.0:/usr/local/share/ruby/gems/2.0
[2017-01-18T18:20:29+00:00] DEBUG: gem_package[bundler] no installed version found for bundler (>= 0)
RubyGems Environment:
  - RUBYGEMS VERSION: 2.0.14.1
  - RUBY VERSION: 2.0.0 (2015-12-16 patchlevel 648) [x86_64-linux]
  - INSTALLATION DIRECTORY: /usr/local/share/ruby/gems/2.0
  - RUBY EXECUTABLE: /usr/bin/ruby2.0
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /usr/local/share/ruby/gems/2.0
     - /root/.gem/ruby/2.0
     - /usr/share/ruby/gems/2.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
[2017-01-18T18:20:29+00:00] DEBUG:  found gem bundler version 1.13.7 for platform ruby from 
[2017-01-18T18:20:29+00:00] DEBUG: gem_package[bundler] bundler not installed, installing 1.13.7
Successfully installed bundler-1.13.7
1 gem installed
[2017-01-18T18:20:31+00:00] INFO: gem_package[bundler] installed bundler at 1.13.7
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[bundler] resetting gem environment to default
[2017-01-18T18:20:31+00:00] INFO: Processing gem_package[io-console] action install (newrelic_plugins::aws_cloudwatch line 53)
[2017-01-18T18:20:31+00:00] DEBUG: Providers for generic gem_package resource enabled on node include: [Chef::Provider::Package::Rubygems]
[2017-01-18T18:20:31+00:00] DEBUG: Provider for action install on resource gem_package[io-console] is Chef::Provider::Package::Rubygems
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] detected omnibus installation in /opt/chef/embedded/bin
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] searching for 'gem' binary in path: /sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin:/opt/chef/embedded/bin:/opt/chef/embedded/bin:/usr/local/sbin:/usr/local/bin
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] using gem '/bin/gem'
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] found installed gem io-console version 0.4.5 matching io-console (>= 0)
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] io-console 0.4.5 already installed
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] is already installed - nothing to do
[2017-01-18T18:20:31+00:00] DEBUG: gem_package[io-console] resetting gem environment to default
`

And results:

`gem list --local

*** LOCAL GEMS ***

bigdecimal (1.2.0)
bundler (1.13.7)
json (1.8.3)
psych (2.0.0)
rdoc (4.2.2)
`

**Comment by lamont-granquist** [comment]
> It appears to be something to do with amazon linux. 

are you all using opsworks?  as brought up in issue #3389 opsworks is running chef-client out of bundler, and then since it needs to escape from the bundle it hacks it up with this monkeypatch:

https://github.com/aws/opsworks-cookbooks/blob/c1426eee3f/opsworks_commons/libraries/monkey_patch_rubygems_provider.rb

**Comment by tophercullen** [comment]
Not using opsworks, I'm using my own chef server with my own cookbooks etc.

**Comment by tophercullen** [comment]
Its been a while, so I had to remember where I had this issue. It was with this: https://github.com/newrelic-platform/newrelic_plugins_chef/blob/master/recipes/aws_cloudwatch.rb

On a new system it still fails for the same root cause.

`  * gem_package[bundler] action install[2017-04-13T18:49:23+00:00] INFO: Processing gem_package[bundler] action install (newrelic_plugins::aws_cloudwatch line 3)
[2017-04-13T18:49:25+00:00] INFO: gem_package[bundler] installed bundler at 1.14.6

    - install version 1.14.6 of package bundler
  * execute[bundle install] action run[2017-04-13T18:49:25+00:00] INFO: Processing execute[bundle install] action run (newrelic_plugins::aws_cloudwatch line 8)

    [execute] /usr/share/ruby/vendor_ruby/2.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- io/console (LoadError)`

**Comment by wjordan** [comment]
My application was affected by this issue, and I think I finally tracked down the root cause.

The issue is that the `AlternateGemEnvironment` class used for searching the system `gem` binary paths for dependencies does so by setting `Gem::Specification.dirs` before invoking the Rubygems API: https://github.com/chef/chef/blob/8a361fbc44402e38b9eac6815c09456211ee0966/lib/chef/provider/package/rubygems.rb#L300

However, setting `dirs` to the alternate-ruby's gem-paths doesn't account for 'default' gems also installed in the embedded Ruby, which still get [added](https://github.com/rubygems/rubygems/blob/c2e957c78f05a67a9ed170ee1ace35c0e39d4646/lib/rubygems/specification.rb#L869) to the list of specifications returned by Rubygems.

Simple fix is to add `.reject(&:default_gem?)` to the end of this line to filter out default gems from the returned list:

https://github.com/chef/chef/blob/8a361fbc44402e38b9eac6815c09456211ee0966/lib/chef/provider/package/rubygems.rb#L90

I tested this fix on my repro and it fixed the issue.

**Comment by lamont-granquist** [comment]
but if the AlternateGemEnvironment has the same gem as the built-in gem in the omnibus ruby, then your solution would incorrectly remove that wouldn't it?

**Comment by wjordan** [comment]
> but if the AlternateGemEnvironment really has the same gem as the built-in gem in the omnibus ruby, then your solution would incorrectly remove that wouldn't it?

Yeah, adding `.reject(&:default_gem?)` would remove the omnibus-ruby 'default' gems from the installed-versions list, which could be considered incorrect in the case where omnibus-ruby is the intended environment to check (and if we intended to include default gems in the installed-versions list). However, the current implementation already incorrectly misses default gems in system ruby environments, so a fix to ignore default gems entirely would at least make the behavior consistent from one environment to the other.

A more conservative fix for this issue would only filter out default gems when they're from the wrong environment:

```
gem_specification.find_all_by_name(gem_dep.name, gem_dep.requirement).reject do |spec|
  # Remove any default gems not from the current environment.
  spec.default_gem? &&
  gem_paths.none? {|dir| spec.loaded_from.include? dir}
end
```

Beyond that, it _might_ be possible to stub more parts of the Rubygems environment beyond `Gem::Specification.dirs` to get the environment-specific default gems added to the installed-versions list as well- but I think that would open up an even greater risk of bugs and issues so I don't think it's worth it.

**Comment by coderanger** [comment]
Should also be fixed by #6955.
